### PR TITLE
ES-498: The volume is not mounted successfully according to nexentastor-block-csi-node Log

### PR DESCRIPTION
**Pull request by Qeas**
Fixed issue that sometimes occured for e2fsck with exit code 8. Moved resize2fs to NodeExpandVolume, removed  e2fsck completely.
Ensure removing all luns before deleting the volume
ES-500: sparse volume support

### CODE REVIEW

**Comment by Qeas** [comment]
Ran 57 of 72 Specs in 281.723 seconds
SUCCESS! -- 57 Passed | 0 Failed | 0 Pending | 15 Skipped
PASS

File: Dockerfile.csi-sanity [hunk 1]
```diff
@@ -68,4 +70,4 @@ RUN /csi-sanity -csi.version &&\
 # other csi-sanity options: --ginkgo.v -ginkgo.noColor
 #TODO remove "--ginkgo.skip" option after fixing volume paginatin by token:
 ENTRYPOINT ["/bin/sh", "-c", "/run-driver && /csi-sanity --csi.endpoint=${SOCK} --csi.createpathcmdtimeout=180 --csi.removepathcmdtimeout=180 --ginkgo.skip 'pagination should detect volumes added between pages and accept tokens when the last volume from a page is deleted'"]
-#ENTRYPOINT ["/bin/sh", "-c", "/run-driver && /csi-sanity --csi.endpoint=${SOCK} --csi.createpathcmdtimeout=180 --csi.removepathcmdtimeout=180]
\ No newline at end of file
+#ENTRYPOINT ["/bin/sh", "-c", "/run-driver && /csi-sanity --csi.endpoint=${SOCK} --csi.createpathcmdtimeout=180 --csi.removepathcmdtimeout=180]
```
Comment (abhishek024kumar) [inline Dockerfile.csi-sanity:73]: minor: Is no newline at end of the file is deliberately removed?

File: examples/kubernetes/nginx-deployment-dynamic-volume.yaml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+# NexentaStor CSI Driver - Persistent Volume Claim
+# ------------------------------------------------
+
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: nexentastor-block-csi-driver-pvc-nginx-dynamic
+spec:
+  storageClassName: nexentastor-block-csi-driver-sc-nginx-dynamic
+  accessModes:
```
Comment (abhishek024kumar) [inline examples/kubernetes/nginx-deployment-dynamic-volume.yaml:27]: For access modes can we please put comments here for users to use only one property at a time just for info(either 'ReadWriteMany' or 'ReadWriteOnce').

File: examples/kubernetes/nginx-dynamic-block-volume.yaml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+# allowedTopologies:
+# - matchLabelExpressions:
+#   - key: topology.kubernetes.io/zone
+#     values:
+#     - zone-1
+parameters:
+  # sparseVolume: "true"               # [optional] Defines whether the volumes should be created sparsed(thin)
+#   csi.storage.k8s.io/fstype: xfs
+#   configName: nstor-box3
+  #dataset: customPool/customDataset # to overwrite "defaultDataset" config property [pool/dataset]
```
Comment (abhishek024kumar) [inline examples/kubernetes/nginx-dynamic-block-volume.yaml:25]: minor: formatting needed.

File: pkg/driver/controllerServer.go [hunk 4]
```diff
@@ -323,16 +325,31 @@ func (s *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi.
...
+    currentSize := volInfo.VolumeSize
+    l.Debugf("Current size of volume %s = %+v", volumePath, currentSize)
+
+    if currentSize < capacityBytes {
+        l.Infof("expanding volume %+v to %+v bytes", volumePath, capacityBytes)
+        err = nsProvider.UpdateVolume(volumePath, ns.UpdateVolumeParams{
+            VolumeSize: capacityBytes,
+        })
+        if err != nil {
+            return nil, fmt.Errorf("Failed to expand volume volume %s: %s", volumePath, err)
```
Comment (abhishek024kumar) [inline pkg/driver/controllerServer.go:343]: typo error: two times volume is printed.

File: pkg/driver/nodeServer.go [hunk 5]
```diff
@@ -926,6 +910,7 @@ func (s *NodeServer) mountVolume(devName, targetPath, fsType string, mountOption
         l.Errorf("Failed to mount device %+v. Error: %v", devName, err)
         return status.Error(codes.Internal, err.Error())
     }
+
```
Comment (abhishek024kumar) [inline pkg/driver/nodeServer.go:913]: minor: extra line, can be removed.

**Review by abhishek024kumar — CHANGES_REQUESTED** [review]
Please address review comments.

File: examples/kubernetes/nginx-dynamic-block-volume.yaml (hunk 3 above)
Comment (Qeas) [reply examples/kubernetes/nginx-dynamic-block-volume.yaml:25]: Could you explain what formatting?

File: examples/kubernetes/nginx-deployment-dynamic-volume.yaml (hunk 2 above)
Comment (Qeas) [reply examples/kubernetes/nginx-deployment-dynamic-volume.yaml:27]: There is also a ReadOnlyMany option. I can add that comment, but those are general kubernetes options, not sure that we need to document all of them.

File: pkg/driver/controllerServer.go (hunk 4 above)
Comment (Qeas) [reply pkg/driver/controllerServer.go:343]: will fix, thanks

File: pkg/driver/nodeServer.go (hunk 5 above)
Comment (Qeas) [reply pkg/driver/nodeServer.go:913]: Although it is optional, I think it is a good visual separation of the "if" block from the rest of the code

File: examples/kubernetes/nginx-deployment-dynamic-volume.yaml (hunk 2 above)
Comment (abhishek024kumar) [reply examples/kubernetes/nginx-deployment-dynamic-volume.yaml:27]: ok, if it's a standard Kubernetes command we can skip it.

File: examples/kubernetes/nginx-dynamic-block-volume.yaml (hunk 3 above)
Comment (abhishek024kumar) [reply examples/kubernetes/nginx-dynamic-block-volume.yaml:25]: space between hash and property.

File: Dockerfile.csi-sanity (hunk 1 above)
Comment (Qeas) [reply Dockerfile.csi-sanity:73]: Yes, I am trying to have a newline at the end of all files and this one was missing it.

**Comment by EvgeniiIvanov** [comment]
Sanity tests -PASS
E2E tests - PASS for K8s 1.19.13-1.20.4-1.21.3
Volume expansion (File System) - PASS for K8s 1.19.13-1.20.4-1.21.3
Volume expansion (Raw Block) - PASS for K8s 1.19.13-1.20.4-1.21.3

**Review by EvgeniiIvanov — APPROVED** [review]
Additional.
E2E tests against K8s 1.22.0 passed.
Expand + CHAP + dynamicTargetLunAllocation - pass.

**Review by goldranjith — APPROVED** [review]
LGTM

**Pull request merged**